prow/jira/jira.go

```diff
@@ -259,8 +288,275 @@ func (jc *client) UpdateRemoteLink(id string, link *jira.RemoteLink) error {
+	result, resp, err := jc.upstream.Issue.AddComment(issueID, comment)
+	if err != nil {
+		return nil, JiraError(resp, err)
+	}
+	return result, nil
+}
+
+func (jc *client) CreateIssue(issue *jira.Issue) (*jira.Issue, error) {
+	result, resp, err := jc.upstream.Issue.Create(issue)
+	if err != nil {
+		return nil, JiraError(resp, err)
+	}
+	return result, nil
+}
+
+func (jc *client) CreateIssueLink(link *jira.IssueLink) error {
+	resp, err := jc.upstream.Issue.AddLink(link)
+	if err != nil {
+		return JiraError(resp, err)
+	}
+	return nil
+}
+
+// GenericCloneIssue is a generic implementation of CloneIssue that can be used with
+// any valid Client.
+func GenericCloneIssue(jc Client, parent *jira.Issue) (*jira.Issue, error) {
+	// create deep copy of parent "Fields" field
+	data, err := json.Marshal(parent.Fields)
+	if err != nil {
+		return nil, err
+	}
+	childIssueFields := &jira.IssueFields{}
+	err = json.Unmarshal(data, childIssueFields)
+	if err != nil {
+		return nil, err
+	}
+	childIssue := &jira.Issue{
+		Fields: childIssueFields,
+	}
+	// update description
+	childIssue.Fields.Description = fmt.Sprintf("This is a clone of issue %s. The following is the description of the original issue: \n---\n%s", parent.Key, parent.Fields.Description)
+
+	// attempt to create the new issue
+	createdIssue, err := jc.CreateIssue(childIssue)
+	if err != nil {
+		// some fields cannot be set on creation; unset them
+		childIssue, err = unsetProblematicFields(childIssue, err)
+		if err != nil {
+			return nil, fmt.Errorf("error identifying unsettable fields during clone creation: %w", err)
+		}
+		createdIssue, err = jc.CreateIssue(childIssue)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// create clone links
+	link := &jira.IssueLink{
+		OutwardIssue: &jira.Issue{ID: parent.ID},
+		InwardIssue:  &jira.Issue{ID: createdIssue.ID},
+		Type: jira.IssueLinkType{
+			Name:    "Cloners",
+			Inward:  "is cloned by",
+			Outward: "clones",
+		},
+	}
+	if err := jc.CreateIssueLink(link); err != nil {
+		return nil, err
+	}
+	// Get updated issue, which would have issue links
+	if clonedIssue, err := jc.GetIssue(createdIssue.ID); err != nil {
+		// still return the originally created child issue here in case of failure to get updated issue
+		return createdIssue, fmt.Errorf("Could not get issue after creating issue links: %w", err)
+	} else {
+		return clonedIssue, nil
+	}
+}
+
+// CloneIssue copies an issue struct, clears unsettable fields, creates a new
+// issue using the updated struct, and then links the new struct as a clone to
+// the original.
+func (jc *client) CloneIssue(parent *jira.Issue) (*jira.Issue, error) {
+	return GenericCloneIssue(jc, parent)
+}
+
+func unsetProblematicFields(issue *jira.Issue, err error) (*jira.Issue, error) {
+	// handle unsettable "unknown" fields
+	// trim create error
+	jsonResponse := strings.TrimPrefix(err.Error(), "request failed. Please analyze the request body for more details. Status code: 400: ")
```

Comment:
Maybe use an error type in `JiraError` that a) keeps the status code so we only do this if we get a 400 back and not on all possible errors and b) makes it easy to access the relevant part of the error?

Reply:
I made a new error type called `JiraError` that stores the response status code and body alongside the original error. I also renamed the `JiraError` function to `HandleJiraError`, which makes more sense to me for a function name.